File: python/0020_valid_parentheses.py
from collections import deque


class Solution:
    def isValid(self, s: str) -> bool:
        if not s:
            return True
        queue = deque()
        for c in s:
            if c == '(':
                queue.append('(')
            elif c == '[':
                queue.append('[')
            elif c == '{':
                queue.append('{')
            elif c == ')':
                if queue:
                    if queue.pop() != '(':
                        return False
                else:
                    return False
            elif c == ']':
                if queue:
                    if queue.pop() != '[':
                        return False
                else:
                    return False
            elif c == '}':
                if queue:
                    if queue.pop() != '{':
                        return False
                else:
                    return False
        if queue:
            return False
        return True


print(Solution().isValid("()[]{}"))
print(Solution().isValid("([]{}"))
print(Solution().isValid("([)]"))

File: cpp/0020_valid_parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        if (s == "") 
            return true;
        if (s.length() % 2 == 1)
            return false;
        stack<char>kuohao;
        for (int i = 0 ; i < s.length(); i++) {
            if (s[i] == '(' || s[i] == '[' || s[i] == '{') {
                kuohao.push(s[i]);
            } else {
                if (kuohao.empty()) {
                    return false;
                }  else if (s[i] == ')' && kuohao.top() == '(') {
                    kuohao.pop();
                } else if (s[i] == '}' && kuohao.top() == '{') {
                    kuohao.pop();
                } else if (s[i] == ']' && kuohao.top() == '[') {
                    kuohao.pop();
                } else {
                    return false;
                }
               
            }
        }
        if (kuohao.empty())
            return true;
        else
            return false;
    }
};
